ZCS-2729 : Scheduler to create contact backups and purge old contact backups

shrikant-prasad opened the pull request:
sending PR on behalf of Chinmay.

ykomiyam commented on store/src/java/com/zimbra/cs/mailbox/ContactBackupThread.java:
```diff
@@ -0,0 +1,396 @@
+    }
+
+    private void createBackup(OperationContext octxt, Mailbox mbox, Account account, SSLConnectionSocketFactory sslcsf, Folder folder, Date startTime) {
+            HttpResponse response = exportContacts(mbox, account, sslcsf);
```
The indent is not correct.  All the contents within the createBackup method should indent-4-space-left.

ykomiyam commented on store/src/java/com/zimbra/cs/mailbox/ContactBackupThread.java:
```diff
@@ -0,0 +1,396 @@
+    private void requestShutdown() {
+        shutdownRequested = true;
+    }
+    // fetches value of zimbraFeatureContactBackupFrequency from server and update the same to a class variable contactBackupFreq
```
fetches => fetch ?  (all other comments use imperative form)

ykomiyam commented on store/src/java/com/zimbra/cs/mailbox/ContactBackupThread.java:
```diff
@@ -0,0 +1,396 @@
+import com.zimbra.cs.util.Config;
+
+public class ContactBackupThread extends Thread{
+    private final static String OPERATION = "ContactBackup";
```
For the integrity of the field declaration, it might be "private **static final**"?

ykomiyam commented on store/src/java/com/zimbra/cs/account/callback/CallbackUtil.java:
```diff
@@ -0,0 +1,140 @@
+            // sanity check, this should not happen because modifyServer is
+            // proxied to the the right server
+            if (server.getId() != localServer.getId())
+                ZimbraLog.misc.warn("%s : wrong server", operation);
```
Please enclose this line with {}

grishick commented:
any idea what's going on with Filter tests on Travis CI?

sneha-patil-synacor commented on store/src/java/com/zimbra/cs/util/Zimbra.java:
```diff
@@ -404,6 +409,7 @@ public static synchronized void shutdown() throws ServiceException {
 
         if (sIsMailboxd) {
             PurgeThread.shutdown();
+            ContactBackupThread.shutdown();
```
Do we need to check ContactBackupThread support here?

shrikant-prasad replied:
yes

sneha-patil-synacor commented on store/src/java/com/zimbra/cs/service/mail/GetContactBackupList.java:
```diff
@@ -32,21 +38,33 @@
-        list.add("file3.tgz");
-        list.add("file4.tgz");
-        return list;
+    private List<String> getContactBackupList(Mailbox mbox) throws ServiceException {
```
Need to update GetContactBackupListTest with this implementation

Prashantsurana commented:
exportContacts method will not work in multi node environment or in the environment where mailbox is not running on 443 port. 

  `  private HttpResponse exportContacts(Account account, SSLConnectionSocketFactory sslcsf)  `

Please consider to avoid writing rest client, instead use zclient or refactor  the export contact code and call API. 

Is there way to set limit on number on contacts per mailbox?

if not this thread can bring down the entire mail store if user creates large number of contacts and everyday backup process will add up load on the mailstore.

shrikant-prasad commented:
1. Modified the contact export logic, now its done using Zmailbox. 
2. added zimbraFeatureContactBackupEnabled ldap attribute to enable/disable the contact backup feature at account level. 
3. Modified the scheduling logic to start the back up as per the backup interval and not to be reset due to server restart.

shrikant-prasad commented on store/src/java/com/zimbra/cs/mailbox/ContactBackupThread.java:
```diff
@@ -0,0 +1,397 @@
+        try {
+            Config.setInt(Config.CONTACT_BACKUP_LAST_MAILBOX_ID, id);
+        } catch (ServiceException se) {
+            ZimbraLog.backup.info("%s : excepion occured while setting contact backup last mailbox id with %d", OPERATION, id);
```
In ContactBackupThread,  'backup' is used for all the logging.

shrikant-prasad commented on store/src/java/com/zimbra/cs/account/callback/CallbackUtil.java:
```diff
@@ -0,0 +1,142 @@
+            Server server = Provisioning.getInstance().getLocalServer();
+            returnValue = server.getTimeInterval(attrName, prevValue);
+        } catch (ServiceException e) {
+            ZimbraLog.purge.warn("Unable to determine value of %s. Using previous value: %d.", attrName, returnValue, e);
```
changed to misc.

Prashantsurana commented:
Please Keep logging consistent.

Backup logger use is ambiguous. Use appropriate logger or introduce new logger. We need to make sure logging should be self explanatory. Anyone should able to distinguish the issue without looking at the source code. 

Print stack trace whereever possible. Only avoid printing the strace where the exceptions are known and harmless.

shrikant-prasad commented on store/src/java/com/zimbra/cs/service/mail/GetContactBackupList.java:
```diff
@@ -32,21 +38,33 @@
+            backup = getContactBackupList(mbox);
+        } catch (ServiceException se) {
+            ZimbraLog.soap.error("Exception occured while fetching list of available contact backups.");
+            throw se;
```
done.

shrikant-prasad commented:
added new logger 'contactbackup' and used that for logging.
